### ISSUE TITLE
share auto loading settings

### ISSUE THREAD

**Issue by fastmatikqc** [oakboat/ApexFreeCheat#2]
**int ur main(); add : cl->checkAndReload(); // Call checkAndReload on the pointer**

struct ConfigLoader {
    const std::string FILE_NAME = "configs.ini";
    std::vector<std::string>* lines = new std::vector<std::string>;
    long m_lastTimeFileEdited = 0;

    // Features
    bool FEATURE_AIMBOT_ON = true;
    bool FEATURE_SENSE_ON = true;
    bool FEATURE_ITEM_GLOW_ON = true;
    bool FEATURE_QUICKTURN_ON = true;
    bool FEATURE_SPECTATOR_ON = true;
    bool FEATURE_SUPER_GLIDE_ON = true;
    bool FEATURE_SKINCHANGER_ON = true;
    bool FEATURE_TRIGGERBOT_ON = true;
    bool FEATURE_NORECOIL_ON = true;
    bool FEATURE_PRINT_LEVELS_ON = true;
    bool FEATURE_MAP_RADAR_ON = true;

    // No Recoil
    int NORECOIL_PITCH_REDUCTION = 15;
    int NORECOIL_YAW_REDUCTION = 17;

    // Trigger Bot
    int TRIGGERBOT_ZOOMED_RANGE = 180;
    int TRIGGERBOT_HIPFIRE_RANGE = 25;

    // Sense
    int SENSE_MAXRANGE = 250;
    int SENSE_MAXRANGE_OVERWALL = 250;

    // Aimbot
    bool AIMBOT_ACTIVATED_BY_ATTACK = true;
    bool AIMBOT_ACTIVATED_BY_ADS = false;
    bool AIMBOT_ACTIVATED_BY_KEY = false;
    float AIMBOT_SMOOTH = 20.211f;
    float AIMBOT_SPEED = 40.211f;
    float AIMBOT_SMOOTH_EXTRA_BY_DISTANCE = 1000.0f;
    float AIMBOT_FOV = 5.0000f;
    bool AIMBOT_PREDICT_BULLETDROP = true;
    bool AIMBOT_PREDICT_MOVEMENT = true;
    bool AIMBOT_ALLOW_TARGET_SWITCH = true;
    int AIMBOT_MAX_DISTANCE = 100;
    int AIMBOT_MIN_DISTANCE = 1;

    // Keys
    std::string AIMBOT_ACTIVATION_KEY = "XK_Shift_L";
    std::string FEATURE_QUICKTURN_BUTTON = "XK_f";
    std::string FEATURE_PRINT_LEVELS_BUTTON = "XK_p";
    std::string FEATURE_MAP_RADAR_BUTTON = "XK_m";
    std::string TRIGGERBOT_PAUSE_BUTTON = "XK_z";

    ConfigLoader() {
        createFileIfNotExists();
        reloadFile();
        std::thread([this]() {
            while (true) {
                std::this_thread::sleep_for(std::chrono::seconds(1)); // Check every second
                checkAndReload(); // Use checkAndReload to see if reload is needed
            }
            }).detach(); // Run in a separate thread
    }

    void createFileIfNotExists() {
        std::ifstream file(FILE_NAME);
        if (!file.good()) {
            std::ofstream outFile(FILE_NAME);
            if (outFile.is_open()) {
                // Optionally write default settings
                outFile << "# Configuration File\n";
                outFile << "FEATURE_AIMBOT_ON=1\n";
                outFile << "FEATURE_SENSE_ON=1\n";
                outFile << "FEATURE_ITEM_GLOW_ON=1\n";
                outFile << "FEATURE_QUICKTURN_ON=1\n";
                outFile << "FEATURE_SPECTATOR_ON=1\n";
                outFile << "FEATURE_SUPER_GLIDE_ON=1\n";
                outFile << "FEATURE_SKINCHANGER_ON=1\n";
                outFile << "FEATURE_TRIGGERBOT_ON=1\n";
                outFile << "FEATURE_NORECOIL_ON=1\n";
                outFile << "FEATURE_PRINT_LEVELS_ON=1\n";
                outFile << "FEATURE_MAP_RADAR_ON=1\n";
                outFile << "NORECOIL_PITCH_REDUCTION=15\n";
                outFile << "NORECOIL_YAW_REDUCTION=17\n";
                outFile << "TRIGGERBOT_ZOOMED_RANGE=180\n";
                outFile << "TRIGGERBOT_HIPFIRE_RANGE=25\n";
                outFile << "SENSE_MAXRANGE=250\n";
                outFile << "SENSE_MAXRANGE_OVERWALL=250\n";
                outFile << "AIMBOT_ACTIVATED_BY_ATTACK=1\n";
                outFile << "AIMBOT_ACTIVATED_BY_ADS=0\n";
                outFile << "AIMBOT_ACTIVATED_BY_KEY=0\n";
                outFile << "AIMBOT_SMOOTH=20.211\n";
                outFile << "AIMBOT_SPEED=40.211\n";
                outFile << "AIMBOT_SMOOTH_EXTRA_BY_DISTANCE=1000.0\n";
                outFile << "AIMBOT_FOV=5.0000\n";
                outFile << "AIMBOT_PREDICT_BULLETDROP=1\n";
                outFile << "AIMBOT_PREDICT_MOVEMENT=1\n";
                outFile << "AIMBOT_ALLOW_TARGET_SWITCH=1\n";
                outFile << "AIMBOT_MAX_DISTANCE=100\n";
                outFile << "AIMBOT_MIN_DISTANCE=1\n";
                outFile.close();
            }
        }
    }

    void reloadFile() {
        if (loadFileIntoVector()) {
            parseLines();
            onSettingsUpdate(); // Notify when settings are updated
        }
    }

    void checkAndReload() {
        struct stat result;
        if (stat(FILE_NAME.c_str(), &result) == 0) {
            long modTime = result.st_mtime;
            if (modTime > m_lastTimeFileEdited) {
                m_lastTimeFileEdited = modTime;
                reloadFile();
            }
        }
    }

    void onSettingsUpdate() {
        std::cout << "Settings updated from file: " << FILE_NAME << std::endl;
    }

    void parseLines() {
        for (const auto& line : *lines) {
            std::vector<std::string> lineParts = split(line);
            if (lineParts.size() != 2) continue; // Skip invalid lines
            // Line key
            std::string key = lineParts[0];
            trim(key);
            if (key.empty()) continue;
            // Line value
            std::string value = lineParts[1];
            trim(value);
            if (value.empty()) continue;
            loadVariables(key, value);
        }
    }

    void trim(std::string& s) {
        ltrim(s);
        rtrim(s);
    }

    void ltrim(std::string& s) {
        s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
            return !std::isspace(ch);
            }));
    }

    void rtrim(std::string& s) {
        s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
            return !std::isspace(ch);
            }).base(), s.end());
    }

    bool loadFileIntoVector() {
        struct stat result;
        if (stat(FILE_NAME.c_str(), &result) == 0) {
            long modTime = result.st_mtime;
            bool fileNeedsReload = modTime > m_lastTimeFileEdited;
            m_lastTimeFileEdited = modTime;
            if (!fileNeedsReload) return false;
        }
        lines->clear();
        std::ifstream myFile(FILE_NAME);
        std::string str;
        while (getline(myFile, str)) {
            trim(str);
            if (str.empty() || str[0] == '#') continue; // Skip empty lines and comments
            lines->push_back(str);
        }
        myFile.close();
        return true;
    }

    void loadVariables(const std::string& key, const std::string& val) {
        // Features
        if (key == "FEATURE_AIMBOT_ON") FEATURE_AIMBOT_ON = toBool(val);
        else if (key == "FEATURE_SENSE_ON") FEATURE_SENSE_ON = toBool(val);
        else if (key == "FEATURE_ITEM_GLOW_ON") FEATURE_ITEM_GLOW_ON = toBool(val);
        else if (key == "FEATURE_QUICKTURN_ON") FEATURE_QUICKTURN_ON = toBool(val);
        else if (key == "FEATURE_SPECTATOR_ON") FEATURE_SPECTATOR_ON = toBool(val);
        else if (key == "FEATURE_SUPER_GLIDE_ON") FEATURE_SUPER_GLIDE_ON = toBool(val);
        else if (key == "FEATURE_SKINCHANGER_ON") FEATURE_SKINCHANGER_ON = toBool(val);
        else if (key == "FEATURE_TRIGGERBOT_ON") FEATURE_TRIGGERBOT_ON = toBool(val);
        else if (key == "FEATURE_NORECOIL_ON") FEATURE_NORECOIL_ON = toBool(val);
        else if (key == "FEATURE_PRINT_LEVELS_ON") FEATURE_PRINT_LEVELS_ON = toBool(val);
        else if (key == "FEATURE_MAP_RADAR_ON") FEATURE_MAP_RADAR_ON = toBool(val);

        // No Recoil
        else if (key == "NORECOIL_PITCH_REDUCTION") NORECOIL_PITCH_REDUCTION = std::stoi(val);
        else if (key == "NORECOIL_YAW_REDUCTION") NORECOIL_YAW_REDUCTION = std::stoi(val);

        // Trigger Bot
        else if (key == "TRIGGERBOT_ZOOMED_RANGE") TRIGGERBOT_ZOOMED_RANGE = std::stoi(val);
        else if (key == "TRIGGERBOT_HIPFIRE_RANGE") TRIGGERBOT_HIPFIRE_RANGE = std::stoi(val);

        // Sense
        else if (key == "SENSE_MAXRANGE") SENSE_MAXRANGE = std::stoi(val);
        else if (key == "SENSE_MAXRANGE_OVERWALL") SENSE_MAXRANGE_OVERWALL = std::stoi(val);

        // Aimbot
        else if (key == "AIMBOT_ACTIVATED_BY_ATTACK") AIMBOT_ACTIVATED_BY_ATTACK = toBool(val);
        else if (key == "AIMBOT_ACTIVATED_BY_ADS") AIMBOT_ACTIVATED_BY_ADS = toBool(val);
        else if (key == "AIMBOT_ACTIVATED_BY_KEY") AIMBOT_ACTIVATED_BY_KEY = toBool(val);
        else if (key == "AIMBOT_SMOOTH") AIMBOT_SMOOTH = std::stof(val);
        else if (key == "AIMBOT_SPEED") AIMBOT_SPEED = std::stof(val);
        else if (key == "AIMBOT_SMOOTH_EXTRA_BY_DISTANCE") AIMBOT_SMOOTH_EXTRA_BY_DISTANCE = std::stof(val);
        else if (key == "AIMBOT_FOV") AIMBOT_FOV = std::stof(val);
        else if (key == "AIMBOT_PREDICT_BULLETDROP") AIMBOT_PREDICT_BULLETDROP = toBool(val);
        else if (key == "AIMBOT_PREDICT_MOVEMENT") AIMBOT_PREDICT_MOVEMENT = toBool(val);
        else if (key == "AIMBOT_ALLOW_TARGET_SWITCH") AIMBOT_ALLOW_TARGET_SWITCH = toBool(val);
        else if (key == "AIMBOT_MAX_DISTANCE") AIMBOT_MAX_DISTANCE = std::stoi(val);
        else if (key == "AIMBOT_MIN_DISTANCE") AIMBOT_MIN_DISTANCE = std::stoi(val);

        // Keys
        else if (key == "AIMBOT_ACTIVATION_KEY") AIMBOT_ACTIVATION_KEY = val;
        else if (key == "FEATURE_QUICKTURN_BUTTON") FEATURE_QUICKTURN_BUTTON = val;
        else if (key == "FEATURE_PRINT_LEVELS_BUTTON") FEATURE_PRINT_LEVELS_BUTTON = val;
        else if (key == "FEATURE_MAP_RADAR_BUTTON") FEATURE_MAP_RADAR_BUTTON = val;
        else if (key == "TRIGGERBOT_PAUSE_BUTTON") TRIGGERBOT_PAUSE_BUTTON = val;
    }

    bool toBool(const std::string& val) {
        return val == "1" || val == "true";
    }

    std::vector<std::string> split(const std::string& s) {
        std::vector<std::string> tokens;
        std::istringstream tokenStream(s);
        std::string token;
        while (std::getline(tokenStream, token, '=')) {
            tokens.push_back(token);
        }
        return tokens;
    }
};
